### ISSUE TITLE
Lint request : avoid_returning_widgets

### ISSUE THREAD

**Issue by NatoBoram** [dart-lang/linter#2410]
### Describe the rule you'd like to see implemented

> Include as much detail as you can. 
> 
> This should include a description of who would use the lint (such as if it's
> specific to a framework like Flutter), what conditions the rule should look
> for, and what conditions the rule should ignore.

Disallow functions from returning a `Widget` except in the case where `@override` is used, such as when building a Widget.

There are many discussions about Functional Widgets vs Stateless Widgets and the consensus is that classes are generally better than functions. I think it would be better to get in the habit of writing widgets instead of functions early in the learning process and a linting rule can help avoid mistakes.

### Examples

> Ideally, provide motivating examples w/ rationale.

* Small stateless widgets have a good chance of being able to use a `const` constructor, which would avoid unnecessary rebuilds.
* Large widgets are easier to maintain when they are in their own file rather than mixed in with another larger widget.

**BAD**

```dart
Widget _getButton() {
  return IconButton(
    tooltip: '',
    icon: null,
    onPressed: () {},
  );
}
```

**GOOD**

```dart
@override
Widget build() {
  return IconButton(
    tooltip: '',
    icon: null,
    onPressed: () {},
  );
}
```

### Additional context
> Add any other considerations or context here.

* https://github.com/flutter/flutter/issues/19269 
* https://flutter.dev/docs/perf/rendering/best-practices#controlling-build-cost

**Comment by incendial** [comment]
If anyone is interested to try this rule, we have it implemented in additional dart linter called [Dart Code Metrics](https://github.com/dart-code-checker/dart-code-metrics/blob/master/doc/rules/avoid-returning-widgets.md). Any feedback is welcome!

**Comment by pq** [comment]
/fyi @dnfield @goderbauer -- we talked about something similar the other day...

**Comment by dnfield** [comment]
I'm failing to see what this adds to address the concerns raised in https://github.com/dart-lang/linter/pull/2582#issuecomment-819728174 and https://github.com/dart-lang/linter/pull/2582#issuecomment-819710459

IOW - there seem to be valid cases where you would return a widget from a function to make code more readable. 

That said, passing around `BuildContext`s to other functions can introduce other issues, particularly if those functions start to think they can hold on to that context (e.g. across an async call boundary or otherwise storing it as a variable).

I'm not strongly for or against this lint.

**Comment by NatoBoram** [comment]
> https://github.com/dart-lang/linter/pull/2582#issuecomment-819728174

Both is good. Such a lint could also have that autofix on save, it would simplify creating new widgets.

> https://github.com/dart-lang/linter/pull/2582#issuecomment-819710459

The point of this lint is to stop doing that and make new widgets.

**Comment by NatoBoram** [comment]
The Flutter team has officially taken a stance in this.

https://youtu.be/IOyq-eTRhvo

**Comment by pq** [comment]
I see the value here but have to defer to flutter folks to get a bit more momentum. 

@goderbauer?

**Comment by NatoBoram** [comment]
Every concern has been addressed years ago, too. Not only that, but there's no an official stance from the Flutter team itself. I don't think there's any reason left to hold back.

**Comment by srawlins** [comment]
Can you summarize the stance from the Flutter team? I apologize, but I'm not going to scour that video for what you are referring to.

**Comment by dleurs** [comment]
@srawlins Flutter in ![this video](https://www.youtube.com/watch?v=IOyq-eTRhvo) recommand creating Statefull / Stateless widget instead of functions that returns a Widget

**Comment by srawlins** [comment]
A recommendation (without knowing any more about it) is very different from a rigorous rule. In order for a rule to be high value (which may lead it to be high priority), it needs to have a very low false positive rate.

If members of the Flutter framework team also would recommend that such a rule be added to the flutter_lints package, meaning it would be of high value, and have very low false positive rates, and the recommendation is very strong, like a developer would have to be in an extreme scenario in order to ever have a function return a Widget, then it could be considered high priority.

Even then, it may not be implemented for many months, as it will be weighed against many other priorities.

**Comment by dleurs** [comment]
It is quite a serious recommendation, even [backed by Remy Rousselet 2 years ago](https://www.reddit.com/r/FlutterDev/comments/mp4xpx/methods_vs_stateless_widget_to_return_widgets/)

"
The definitive answer is to use classes over functions.
That is what the Flutter team and most experts will recommend.
"

I think this request fits well inside an optionnal linter rule. Most linter rules are not following rigorous rule, and I think this request improve perfomances significantly

**Comment by NatoBoram** [comment]
Here's a transcript for people in a hurry.

> First and foremost, remember that when `setState` is called within a widget, its entire build method is rerun. This means that if a user toggles a favorite icon in the corner of a large piece of your UI, having rendered that icon in a helper method will require Flutter to rebuild the entire wrapping widget.
> 
> You may have heard that there are multiple trees that Flutter uses to build your UIs, and the widget tree is only the top one. Creating a few more widgets in that top tree is hardly ever going to slow down your app, but unnecessarily rebuilding whole sections of your UI can cause those deeper trees: the element tree and the render object tree to waste precious CPU time. And what's worse, consider what happens if you animate your icon between states. Now your app is unnecessarily redrawing expensive UI chunks 60 times per second for the duration of the animation, when all it had to do was animate the pixels inside the icon.
> 
> So, instead of refactoring that favorite icon into a helper method, use an entirely new widget up front, so that when it toggles states, Flutter is able to precisely target what it re-renders. And for truly best results, use const constructors wherever possible as that allows Flutter to short-circuit the most amount of unnecessary work.
> 
> Congrats! Now your app is running faster and everyone's happy.
> 
> It's time to write some tests for your UI. So you open up that widgets file again to remind yourself what to capture in a widget test. One obvious candidate is that the favorite icon we've been talking about: changes color as expected.
> 
> Which piece of code would you rather write a test for? The helper method version? Or the separate widget version? The separate widget version couldn't be simpler, whereas the helper method version requires you to reconstruct every dependency needed by `_expensiveWidget1()`, and `_expensiveWidget2()`, and they were expensive, so it's probably a lot. As if that's not enough, there are other scenarios where you can accidentally hold on to a stale `BuildContext`.
> 
> Consider this code which uses a `Builder`, but accidentally uses a different name for one of the `builtContexts`, allowing further nested code to use an old, unreliable `builtContexts`. Refactoring this into a separate widget can make the bug impossible. Now our `MyIconButton` only has access to one `BuiltContext`, and it's always the correct one. Yes, it's true that creating new widgets severs that free connection back to all of the available attributes.
> 
> Except, now that we've explored it more, it never really was free, was it? It always came at the cost of performance, testability and occasionally, even accuracy. When I was speaking about this issue with Remi Rousselet, the creator of Provider, Riverpod, Freezed, and many other packages, he summarized a situation like this:
> 
> > "Classes have a better default behavior. the only benefit of methods is having to write a tiny bit less code. There's no functional benefit."
> 
> Thanks Remi, for helping draw attention to this issue with some instructive Dart Pads.
> 
> And that's it, folks. Next time you're breaking up an unwieldy build method, don't be afraid of specialized widgets. You pressing a few more keys during development, might just save your users dropped frames during runtime.

**Comment by goderbauer** [comment]
What Dan said in https://github.com/dart-lang/linter/issues/2410#issuecomment-940480320 and what is said in the comments linked from there is still accurate. While in general it is a good idea to prefer widget classes over functions it is not a hard rule. In fact, there are many cases where one has no other choice, but to define a function that returns a widget (for example to be passed to Builder.builder or Builder-like widgets). There is nothing wrong with that and that is by design.

I don't see us adding this lint to flutter_lints or recommending it in any other way.

**Comment by NatoBoram** [comment]
> In fact, there are many cases where one has no other choice, but to define a function that returns a widget (for example to be passed to Builder.builder or Builder-like widgets).

Can you find other cases than the ones noted in the proposal? There's also code examples in the proposal that can help you.

> Disallow functions from returning a `Widget` except in a few exceptions.
> 
> * `@override` is used
> * It's a type, not an implementation
> * It's an anonymous function // <- Here's your Builder use case

**Comment by JaredEzz** [comment]
+1 

I'd like to be able to add this to my `analysis_options.dart` to get warnings when returning Widgets from methods because it interrupts the build() chain causing problems with Bloc state management. It would be helpful for newer members of my team who are still unfamiliar with some Flutter conventions.
